torch/_dynamo/variables/user_defined.py

```diff
@@ -510,3 +510,28 @@ def odict_getitem(self, tx, key):
         ).add_options(
             key, self
         )
+
+
+class ProcessGroupVariable(UserDefinedObjectVariable):
+    """
+    We don't want a ProcessGroup object to end up in our output graph.
+
+    But it's common for dynamo to intercept a PG that is then used to get info like
+    rank() or world_size(), as well as passed to utility functions in distributed_c10d
+    which desugar it into plain types like a ranklist and tag.
+
+    For convenience and proper guarding, we construct a variable type.
+
+    TODO: implement+test common/safe methods such as get_rank
+    TODO: make it possible to use ProcessGroupVariable as input to simple functions
+          like _expand_group without dynamo complaining about making a proxy for it.
+          It is not a tensor-like type, and we don't want a proxy- but dynamo assumes
+          torch library functions are dealing with tensor-like types and would have proxies
+          for their args.
+    """
+
+    def __init__(self, value, **kwargs):
+        super().__init__(value, **kwargs)
+
+    def as_python_constant(self):
+        return self.value
```

Comment:
Rewritten from the top-down fsdp tracing experimentation:

```suggestion
class ProcessGroupVariable(UserDefinedObjectVariable):
    """
    We don't want a ProcessGroup object to end up in our output graph.
    But it's common for dynamo to intercept a PG that is then used to get info like
    rank() or world_size(), as well as passed to utility functions in distributed_c10d
    which desugar it into plain types like a ranklist and tag.
    For convenience and proper guarding, we construct a variable type.
    TODO: implement+test common/safe methods such as get_rank
    TODO: make it possible to use ProcessGroupVariable as input to simple functions
          like _expand_group without dynamo complaining about making a proxy for it.
          It is not a tensor-like type, and we don't want a proxy- but dynamo assumes
          torch library functions are dealing with tensor-like types and would have proxies
          for their args.
    """

    def __init__(self, value, **kwargs):
        super().__init__(value, **kwargs)

    def as_python_constant(self):
        return self.value
    
    def call_function(
        self, tx, args: "List[VariableTracker]", kwargs: "Dict[str, VariableTracker]"
    ) -> "VariableTracker":
        print("PG INVOKED", self, [type(arg) for arg in args])
        return super().call_function(tx, args, kwargs)

    def call_method(
        self,
        tx,
        name,
        args: "List[VariableTracker]",
        kwargs: "Dict[str, VariableTracker]",
    ) -> "VariableTracker":
        if name == "rank":
            return variables.ConstantVariable(self.value.rank)
        if name == "size":
            return variables.ConstantVariable(self.value.rank)
        return super().call_method(tx, name, args, kwargs)
        
    def var_getattr(self, tx, name):
        if name in ["rank", "size"]:
            return variables.LambdaVariable(
                lambda *args, **kwargs: self.call_method(tx, name, args, kwargs)
            ).add_options(self)
        return super().var_getattr(tx, name)
```

Reply:
Ahh LambdaVariable is great. I didn't know we had that.

Also, I don't understand why we want `call_function` - is it ever valid to write user code like `my_processgroup()`?  (Maybe it is, but i just didn't think that was a thing)

Reply:
You can drop the `call_function` - this is just here for debugging. It will automatically go to super on its own.

Reply:
cool. yea i dropped it